docs/Hardening.md

```diff
@@ -0,0 +1,197 @@
+#### Issue
+Several major folders in Linux need to have appropriate file permissions, otherwise the OMS Agent won't run properly. Specifically, we need the following:
+| Directory | Permissions (numeric value) | Permissions (symbolic notation) |
+| --- | --- | --- |
+| `/var` | 755 | `drwxr-xr-x` |
+| `/var/log` | 775 | `drwxrwxr-x` |
+| `/etc` | 775 | `drwxrwxr-x` |
+| `/tmp` | 775 | `drwxrwxr-x` |
+
+#### How to Check
+You can check what the permissions for these folders are by the following:
+```
+$> ls -l / | grep var
+dr-x--x--x  14 root root       4096 Apr  6  2019 var
+$> ls -l /var | grep log
+dr-x--x--x  11 root syslog     4096 Nov 17 00:05 log
+$> ls -l / | grep etc
+dr-x--x--x 132 root root      12288 Nov 14 06:30 etc
+$> ls -l / | grep tmp
+dr-x------  20 root root       4096 Nov 17 08:38 tmp
+```
+
+#### How to Fix
+If you find any of the above directories have too strict permissions, you can change the permissions by the following:
+```
+$> chmod 755 /var
+$> chmod 775 /var/log
+$> chmod 775 /etc
+$> chmod 775 /tmp
+```
+
+### Hard Limitation to Thread Resources
+#### Issue
+It's possible to set a security limitation to limit the number of threads any process running as omsagent user (like omsagent, dsc, omi, npm, ...) to create a maximum which is too low for the agent to run successfully. The omsagent user needs to be able to run at least 200 threads in order to run successfully.
+
+#### How to Check
+You can see the number of threads omsagent is limited to by running:
+```
+$> cat /etc/security/limits.conf | grep omsagent
+omsagent  hard  nproc  75
+```
+**Note:** If you don't see anything upon running the above command, it means that omsagent isn't being limited, and this isn't the problem.
+
+It's also possible that this is the issue if you see the below log show up in omsagent.log:
+```
+2020-01-16 10:38:57 -0800 [error]: unexpected error error_class=ThreadError error=#<ThreadError: can't create Thread: Resource temporarily unavailable>
+```
+
+#### How to Fix
+If you increase the number to 200, then that should solve the issue immediately. This can be done by opening up `/etc/security/limits.conf` in a text editor, and changing the number next to omsagent to 200.
+
+You can check if it was successful by doing the following:
+```
+$> cat /etc/security/limits.conf | grep omsagent
+omsagent  hard  nproc  200
+```
+
+### Existing Ruby from RVM
+#### Issue
+Because the OMS Agent ships a specific version of Ruby, there's a possibility of installing a different version of Ruby from the RVM (Ruby version manager). This can end up with conflicts with the OMS version, due to the global variables set by RVM (ex: GEM_HOME, GEM_PATH).
+
+#### How to Check
+You can check the version of Ruby that's running with the following commands:
+```
+$> ruby --version
+ruby 2.2.2p95 (2015-04-13 revision 50295) [x86_64-linux]
+$> which ruby
+/usr/local/rvm/rubies/ruby-2.2.2/bin/ruby
+```
+Standard Ruby install through OMS Agent should result in below:
```

Comment:
This confused me personally; my understanding was that most machines will have _system_ Ruby installed at `/usr/bin/ruby`, while we (OMS) install our own ruby at `/opt/microsoft/omsagent/ruby/bin/ruby`. This section seems to imply that we install the OMS-shipped ruby at `/usr/bin/ruby`, which is not the case. 

Assuming my understanding is correct (@abenbachir ?) maybe the line can be changed to:
"The standard Ruby install should result in the below:" 

And the "Issue" description might want to change to:
"OMS Agent ships a specific version of Ruby. Installing a different version of Ruby from the RVM (Ruby Version Manager) may define global envvar (e.g. `GEM_HOME`, `GEM_PATH`) that conflict with those used by the OMS-specific Ruby installation."

Reply:
@hestolz maybe I haven't explain it more clear, but the Official Ruby shipped by any linux distribution is not the issue here, but rather the custom Ruby installation that is made through RVM. Feel free to improve the wording.

Reply:
Clarified Ruby installation wording:
https://github.com/microsoft/OMS-Agent-for-Linux/pull/1269/commits/b6a96c6bef0fd670a3c6676c17026d5228135af5#diff-ff8800189743d6567bcda8a83de5b22c943b28c51e4488cce0ea2b185716cc37R90